### PR TITLE
[nametab] Introduce type of imperative nametabs, unify API

### PR DESCRIPTION
**Pull request by ejgallego**
This is both a cleanup and a step towards pushing the state upwards, and handling the nametabs functionally.

Related to https://github.com/coq/coq/pull/16746 and https://github.com/coq/ceps/pull/65

Overlays:

### CODE REVIEW

File: library/nametab.mli [hunk 1]
```diff
@@ -104,46 +107,73 @@ type visibility = Until of int | Exactly of int
...
+  val locate_all : qualid -> elt list
+  val exists : path -> bool
+  val path : elt -> path
+
+  (* future work *)
+  (* val shortest_qualid : *)
+end
 
-val push_universe : visibility -> full_path -> Univ.UGlobal.t -> unit
+(* We will introduce a [Make] functor and remove NAMETREE below when libobject and interact with the Nametab *)
```
Comment (SkySkimmer) [inline library/nametab.mli:126]: "when libobject and interact" does not make sense

**Review by SkySkimmer — COMMENTED** [review]
Why the new module type S instead of using nametree?

**Comment by ejgallego** [comment]
> Why the new module type S instead of using nametree?

Because `S` is the imperative API, `NAMETREE` the functional one. Maybe we can get rid of `S`, but we need to see.

Sorry for not marking this a draft by the way.

**Comment by gares** [comment]
Would you mind separating commits doing just alpha renamig so that the pr us easy to review?

**Comment by ejgallego** [comment]
> Would you mind separating commits doing just alpha renaming so that the pr is easy to review?

I don't, tho I'm unsure what can I separate more; do you have an example?

**Comment by SkySkimmer** [comment]
I think comments from november got confused somehow into thinking this PR does more than it does. AFAICT it is mostly alpha renaming.

Is there a reason to keep the PR as draft?

**Comment by ejgallego** [comment]
> I think comments from november got confused somehow into thinking this PR does more than it does. AFAICT it is mostly alpha renaming.

Yes, the PR is mostly redoing the API into a module, plus a couple of very minor cleanups that should not be behavior-changing. But the idea of this PR is not to alter behavior, but allow subsequent PRs to profit from the new uniform API (so objects declaration can be functors over nametab modules).

> Is there a reason to keep the PR as draft?

- I'm not sure I have converged yet on the API
- The overlays needs preparation

But indeed it seems to me the PR is close to be unmarked as draft.

**Pull request closed without merging**